### ISSUE TITLE
Automapping: Wildcard input index

### ISSUE THREAD

**Issue by eishiya** [mapeditor/tiled#3636]
tl;dr version: Wildcard input indices, allowing certain input layers to be shared among all the input indices of a rule.

A couple of times I've seen people get confused about input indices in a reasonable way: they think that if they have an input "without" an index, it'll be part of all indices.

For example:
> input1_Detail
> input2_Detail
> input_Ground

They think input_Ground will behave like two identical input1_Ground and input2_Ground layers.

Of course, the truth is that an empty index is just another index. However, it would be quite convenient if it was possible to mark certain input layers as being used for all indices, using a wildcard index.
It's fairly common in my rules that address multiple layers that only inputs from one layer need to have independent indices, and inputs from other layers are the same for all indices, necessitating a bunch of duplication. A wildcard index would avoid this.

While I think the usefulness of this is clear enough, what the wildcard should *look* like is less clear:
- `*`, e.g. `input*_foo`: this is widely used for wildcards already, and it's not likely anyone is using this as an actual index already. However, people may expect this to perform partial string matching, they might expect `inputA*` to match `inputAbove`, for example.
- Empty index, e.g. `input_foo`: this would probably be intuitive to new users as the fact that an empty index is an actual separate index seems to confuse folks. But, existing Automapping rules may include combinations of the empty index and named indices and expect them to all be separate indices. I know I have some rules like that.
- some reserved keyword, such as `ALL`, e.g. `inputALL_foo`: this would avoid the drawbacks of the above approaches and isn't likely anyone's using something like "ALL" as an index already, but it's longer and not language-neutral - but neither is "input", of course.

I'd love to hear people's preferences on different ways to do the wildcard.

**Comment by eishiya** [comment]
Now that `output`s with an empty index are always output as of #3909, perhaps a similar change could be made for `input`? Changing the meaning of an empty input index may break existing rules, but probably not much more than the output change did, and it would at least make for consistent/comparable behaviour for empty indices.

**Comment by AndreaBellocci** [comment]
Hi, is this issue still relevant? I'd like to try to address it

**Comment by bjorn** [comment]
@AndreaBellocci Yes, this issue is still relevant. You can have a look at #3909 because it made a related change to the output index, though right now I'm not sure if that helps figure out the implementation for the same behavior of the input index.

In any case the decision is not to use an explicit wildcard like `*` in the index, but rather change the behavior for the empty index (`input_[layer-name]`), the second option suggested above by eishiya.

It's already explained quite well above, but for completeness sake, say you have these input layers:

* `input_foo`
* `input1_foo`
* `input2_foo`

It is currently treated by Tiled as three input indexes, and rules using these layers will apply when any of them matches the input. With the new behavior, the above should behave such that the part in `input_foo` is always required to match, in addition to one of the non-empty input indices (unless they're both without content).

Btw, if you're not familiar with [Automapping](https://docs.mapeditor.org/en/stable/manual/automapping/) it should pay off to read up on it and play around with it a bit in Tiled before diving straight into the code.

Also, there are quite elaborate autotests for Automapping in `tests/automapping`. Though they're not set up in the ideal way, requiring the working directory to be `tests/automapping`, so running them from the CLI becomes for example:

```
bjorn@thor:tiled$ cd tests/automapping/
bjorn@thor:automapping$ qbs run -f ../../tiled.qbs -d ../../default -p test_automapping 
```

You can also run them from Qt Creator, if you add a run configuration for them and set the correct working directory. You'll definitely want to add a test for this new behavior, or expand an existing test when appropriate.

I'll assign the issue to you. Just let me know when you change your mind about it!

**Comment by AndreaBellocci** [comment]
Hi, thank you. I spent yesterday understanding the automapper code and I think I get it now, but I have a question. Should theunindexed layer be used only as a required layer, or can it exist on its own? I mean, if the unindexed layer has some matches but then no other layer can be applied, should the output be produced anyway?

**Comment by eishiya** [comment]
Wildcard matching should be equivalent to duplicating that layer to every named index. This means that if a wildcard matches but no named index produces a match, then there is no match - unless, of course, the rules map contains no named indices.

**Comment by AndreaBellocci** [comment]
And if the rules map does not contain any named indeces, the wildcard should just be treated as a normal input, right?

**Comment by eishiya** [comment]
My explanation above is probably misleading. The important part is "wildcard matching should be equivalent to duplicating that layer to every named index." Beyond that, I can't really answer your (previous) question because what that means depends on what exactly we're looking at. So, some examples, which may also be useful in case you don't fully understand how Automapping matching currently works.
Let's say we're potentially looking at "Details" and "Ground" layers. When I speak of match/no match below, that is *for a given cell*, not necessarily the entire rule. In situations where an index has multiple input layers looking at the same working layer, multiple partial matches can add up to an overall match, and that doesn't lend itself well to a simple text-based example like this.

First, some examples that all look at the same working layer:

> inputA_Ground - match
> inputB_Ground - no match
> input_Ground - no match

The above **matches** because the Ground layer was matched on index A. The wildcard just adds an alternative option to the Ground cell to both indices.

> inputA_Ground - no match
> inputB_Ground - no match
> input_Ground - match

The above also **matches** on A and B, for the same reason: the wildcard has added whatever is on it as an option to both A and B.

> inputA_Ground - no match
> inputB_Ground - match
> input_Ground - match

The above **matches** on both A and B.

> inputA_Ground - no match
> inputB_Ground - no match
> input_Ground - no match
> input_Ground - match

The above **matches** on A and B just like the above example. In this example, we have two wildcard Ground layers, which means we're adding two match options to A and B.

In the above four examples, a match anywhere makes for a match overall, because the wildcard just adds another option (or multiple other options) that counts as a match to the Ground layer.

> inputA_Ground - no match
> inputA_Ground - match
> inputB_Ground - no match
> input_Ground - no match

This example is just like the first example. We **match** on index A.

Things get more interesting when we start looking at multiple working layers:

> inputA_Ground - no match
> inputB_Details - match
> input_Ground - match

The above is a match, and *both A and B are matches*. A matches because it has two ways for Ground to match and one of them (the wildcard) matched, and Ground is the only layer A cares about. B matches because both Details (via named index) and Ground (via wildcard) matched.

> inputA_Ground - no match
> inputA_Details - match
> inputB_Ground - no match
> input_Ground - match

The above **matches** on A and B. A matches because the wildcard provides a Ground match, and index A provides a Details match. B matches because the wildcard provides a Ground match, and B does not check any other layers.

> inputA_Ground - no match
> inputA_Details - no match
> inputB_Details - no match
> input_Ground - match

The above is **not a match**. A matches on Ground (via wildcard) but there's no match for Details. B also matches on Ground via the wildcard, but has no match for Details.

> inputA_Ground - no match
> inputA_Details - no match
> inputB_Details - match
> input_Ground - no match

The above is also **not a match**. Both A and B look at both Ground and Details, but there is no Ground match anywhere.

> inputA_Details - match
> inputB_Ground - match
> inputB_Details - match
> input_Ground - no match

This one **matches** on B. A doesn't match because while Details matches, the requirement for the Ground layer added by the wildcard means it needs to match Details *and* Ground, but there is no Ground match.

Hopefully the examples clarify the situation somewhat rather than just confusing you further xP
Again, pretend like any wildcards are copied into every index. That's how situations like this must be handled in Automapping currently.

---
If a map contains no named indices, then the wildcards should operate as their own "empty" index, yes. Theoretically this should not require special handling, since the wildcards just add options (and layer requirements). But I haven't looked at the Automapping code enough to be sure.

**Comment by eishiya** [comment]
For completeness' sake, an example with wildcards looking at two layers at once, though it doesn't really introduce any concepts beyond the previous examples:

> inputA_Ground - match
> inputB_Details - match
> input_Ground - no match
> input_Details - no match
> input_Details - match

This one **matches**  on A. The wildcards add a requirement to look at Ground and Details, and the indexed A layer provides a Ground match, while a wildcard provides a Details match. B doesn't match because the wildcards add a requirement for Ground, but there's no Ground match anywhere.

I'm off to bed now, so if I've confused you, I'm afraid any clarification from me will have to wait a while.

**Comment by AndreaBellocci** [comment]
Ok I think I've completed it. It should match the behaviour that you requested. What should the PR contain? (I have never contributed to open source softwares before so I'm not sure how to do it properly, sorry). I also have some comments on the automapping gui and possible improvements, where can I discuss them?
Thank you,
Andrea

**Comment by eishiya** [comment]
The PR should contain the code changes necessary for this Automapping change, and no others (so if you made some other tweaks on your branch, those should not be in the PR).
Include a line in `NEWS.md` that briefly describes the change, e.g. "AutoMapping: Input layers with no index now act as wildcards, part of every index."
Ideally, you should also include changes to the manual's AutoMapping page reflecting this new behaviour. If you're not able to write the relevant docs yourself, mention in your PR description that you'd like someone else to add documentation for it. If documentation goes unmentioned, it's more likely to be forgotten xP

@bjorn Should we somehow emphasise that this is a compatibility-breaking change? It'll at least be worth a mention in the release blog post. IMHO `NEWS.md` should mark potentially breaking changes since that's the list people looking for a source of incompatibilities will be searching rather than blog posts, because it contains everything, but it doesn't look like compatibility-breaking changes are currently marked in any way.

**Comment by bjorn** [comment]
@eishiya Yeah I think it's better to mention compatibility breaking changes in the release post rather than `NEWS.md`, since that file is rather hidden anyway. I think it's also worth a mention in the manual.

Also, because it breaks compatibility, I guess we can do the usual thing where we'll not do this new behavior in "legacy" mode, because we don't expect people to define rules using region layers and also rely on this feature. Though of course the period between Tiled 1.9 and today keeps growing, which increases the chance that people are affected by this that have already written their rules without region layers, heh.

**Comment by AndreaBellocci** [comment]
Hi, i just wanted to confirm that I've kept legacy mode as it was.
